File: node/etsm/test/virtual_call.js
var etsm = require('./../etsm');
var assert = require('assert');

class FooState extends etsm.State {
    constructor(enter, exit, tick) {
        super(enter, exit);
        this.Tick = tick;
    }
}

class Foo {
    constructor() {
        this.sm = new etsm.StateMachine(this);
        this.a = new FooState(null, null, this.TickA);
        this.b = new FooState(null, null, this.TickB);
        this.output = '';
    }

    TickA(){
        this.output += ' A ';
    }

    TickB(){
        this.output += ' B ';
    }

    Tick(){
        if ( this.sm.current != null && this.sm.current.Tick != null)
            this.sm.current.Tick.apply(this);
    }

    Test() {
        assert(this.sm.IsIn(null));

        this.sm.Transition(this.a);
        assert(this.sm.IsIn(this.a));
        this.Tick();

        this.sm.Transition(this.b);
        assert(this.sm.IsIn(this.b));
        this.Tick();

        this.sm.Transition(null);
        assert(this.sm.IsIn(null));

        assert(this.output == ' A  B ');
    }    
}

foo = new Foo();
foo.Test();

File: ruby/etsm/test/virtual_call.rb
require '../lib/etsm.rb'

class FooState < Etsm::State
    attr_reader :tick

    def initialize(enter, exit, tick)
        super(enter, exit)
        @tick = tick
    end
end

class Foo
    def initialize
        @sm = Etsm::StateMachine.new(self)
        @a = FooState.new(nil, nil, self.method(:TickA))
        @b = FooState.new(nil, nil, self.method(:TickB))
        @output = ''
    end

    def TickA
        @output += ' A '
    end

    def TickB
        @output += ' B '
    end

    def Tick
        if @sm.current != nil and @sm.current.tick != nil
            @sm.current.tick.call
        end
    end

    def Test()
        raise "state != nil" unless @sm.IsIn(nil)
        self.Tick

        @sm.Transition(@a)
        raise "state != A" unless @sm.IsIn(@a)
        self.Tick

        @sm.Transition(@b)
        raise "state != B" unless @sm.IsIn(@b)
        self.Tick

        @sm.Transition(nil)
        raise "state != nil" unless @sm.IsIn(nil)
        self.Tick

        raise 'Output != " A  B "' unless @output == ' A  B '
    end
end

foo = Foo.new
foo.Test()

